gctscript/vm/vm.go

```diff
@@ -0,0 +1,242 @@
+	code, err := ioutil.ReadAll(f)
+	if err != nil {
+		return &Error{
+			Action: "Load: Read",
+			Script: file,
+			Cause:  err,
+		}
+	}
+
+	vm.File = f.Name()
+	vm.Path = filepath.Dir(file)
+	vm.Script = script.New(code)
+	vm.Script.SetImports(loader.GetModuleMap())
+
+	if GCTScriptConfig.AllowImports {
+		if GCTScriptConfig.Verbose {
+			log.Debugf(log.GCTScriptMgr, "File imports enabled for vm: %v", vm.ID)
+		}
+		vm.Script.EnableFileImport(true)
+	}
+	vm.event(StatusSuccess, "load", true)
+	return nil
+}
+
+// Compile compiles to byte code loaded copy of vm script
+func (vm *VM) Compile() (err error) {
+	vm.Compiled = new(script.Compiled)
+	vm.Compiled, err = vm.Script.Compile()
+	return
+}
+
+// Run runs byte code
+func (vm *VM) Run() (err error) {
+	if GCTScriptConfig.Verbose {
+		log.Debugf(log.GCTScriptMgr, "Running script: %s ID: %v", vm.ShortName(), vm.ID)
+	}
+
+	err = vm.Compiled.Run()
+	if err != nil {
+		vm.event(StatusFailure, TypeExecute, true)
+		return Error{
+			Action: "Run",
+			Cause:  err,
+		}
+	}
+	vm.event(StatusSuccess, TypeExecute, true)
+	return
+}
+
+// RunCtx runs compiled byte code with context.Context support.
+func (vm *VM) RunCtx() (err error) {
+	if vm.ctx == nil {
+		vm.ctx = context.Background()
+	}
+
+	ct, cancel := context.WithTimeout(vm.ctx, GCTScriptConfig.ScriptTimeout)
+	defer cancel()
+
+	if GCTScriptConfig.Verbose {
+		log.Debugf(log.GCTScriptMgr, "Running script: %s ID: %v", vm.ShortName(), vm.ID)
+	}
+
+	err = vm.Compiled.RunContext(ct)
+	if err != nil {
+		vm.event(StatusFailure, TypeExecute, true)
+		return Error{
+			Action: "RunCtx",
+			Cause:  err,
+		}
+	}
+	vm.event(StatusSuccess, TypeExecute, true)
+	return
+}
+
+// CompileAndRun Compile and Run script with support for task running
+func (vm *VM) CompileAndRun() {
+	err := vm.Compile()
+	if err != nil {
+		log.Error(log.GCTScriptMgr, err)
+		err = RemoveVM(vm.ID)
+		if err != nil {
+			log.Error(log.GCTScriptMgr, err)
+		}
+		return
+	}
+
+	err = vm.RunCtx()
+	if err != nil {
+		log.Error(log.GCTScriptMgr, err)
+		err = RemoveVM(vm.ID)
+		if err != nil {
+			log.Error(log.GCTScriptMgr, err)
+		}
+		return
+	}
+
+	if vm.Compiled.Get("timer").String() != "" {
+		vm.T, err = time.ParseDuration(vm.Compiled.Get("timer").String())
+
+		if err != nil {
+			log.Error(log.GCTScriptMgr, err)
+			err = vm.Shutdown()
+			if err != nil {
+				log.Error(log.GCTScriptMgr, err)
+			}
+			return
+		}
+		if vm.T < time.Nanosecond {
+			log.Error(log.GCTScriptMgr, "Repeat timer cannot be under 1 nano second")
+			return
+		}
+		vm.runner()
+	} else {
+		err = vm.Shutdown()
+		if err != nil {
+			log.Error(log.GCTScriptMgr, err)
+		}
+		return
+	}
+}
+
+// Shutdown shuts down current VMP
+func (vm *VM) Shutdown() error {
+	if vm.S != nil {
+		vm.S <- struct{}{}
+		close(vm.S)
+	}
+	if GCTScriptConfig.Verbose {
+		log.Debugf(log.GCTScriptMgr, "Shutting down script: %s ID: %v", vm.ShortName(), vm.ID)
+	}
+	vm.Script = nil
+	pool.Put(vm.Script)
```

Comment:
This is to clear out vm.Script _before_ putting it back into the pool for reuse 
while its not a huge requirement it helps stop any potential information leaks and also GC

Reply:
If we look at the pool.Put func it checks the interface if == nil and just returns. I would suggest getting rid of the pool entirely because the the underlying script fields are unexported and can't be cleared for reuse.

Reply:
Things you learn nvm. 🥇